jbmc/unit/java_bytecode/java_bytecode_instrument/virtual_call_null_checks.cpp

```diff
@@ -0,0 +1,94 @@
+/*******************************************************************\
+
+ Module: Unit test to check Java virtual calls via a pointer
+         yield a correct sequence of not-null assumptions.
+
+ Author: Diffblue Limited.
+
+\*******************************************************************/
+
+#include <testing-utils/catch.hpp>
+#include <testing-utils/message.h>
+#include <java-testing-utils/load_java_class.h>
+
+#include <analyses/local_safe_pointers.h>
+#include <goto-programs/goto_convert_functions.h>
+#include <java_bytecode/java_bytecode_language.h>
+#include <util/expr_iterator.h>
+
+SCENARIO(
+  "java_bytecode_virtual_call_null_checks",
+  "[core][java_bytecode][java_bytecode_instrument]")
+{
+  GIVEN("A class that makes a virtual call via a pointer")
+  {
+    symbol_tablet symbol_table = load_java_class(
+      "VirtualCallNullChecks", "./java_bytecode/java_bytecode_instrument");
+    namespacet ns(symbol_table);
+    goto_functionst goto_functions;
+    goto_convert(symbol_table, goto_functions, null_message_handler);
+
+    WHEN("The virtual call is converted")
+    {
+      const auto &main_function =
+        goto_functions.function_map.at("java::VirtualCallNullChecks.main:()V");
+
+      THEN("The callee should be of the form some_ptr->field")
+      {
+        bool found_virtual_call = false;
+
+        for(const auto &instruction : main_function.body.instructions)
+        {
+          if(instruction.type == FUNCTION_CALL)
+          {
+            const auto &virtual_call = to_code_function_call(instruction.code);
+            if(virtual_call.function().id() == ID_virtual_function)
+            {
+              REQUIRE(virtual_call.arguments().size() == 1);
+              const auto &this_argument = virtual_call.arguments()[0];
+              REQUIRE(this_argument.id() == ID_member);
+              REQUIRE(this_argument.op0().id() == ID_dereference);
+              found_virtual_call = true;
+            }
+          }
+        }
+
+        REQUIRE(found_virtual_call);
+      }
+      THEN("All pointer usages should be safe")
+      {
+        local_safe_pointerst safe_pointers(ns);
+        safe_pointers(main_function.body);
```

Comment:
I'm not sure it is best practise for unit test to depend on an entirely separate analysis (it seems as likely as not that this analysis would miss that `ASSUME(a->b != null && a != null) a->b.function()` TBH. I'd probably explicitly assert what you're expecting in the code

Reply:
It's tricky -- I want to assert that every use of a pointer is preceded by a corresponding ptr != null assumption. I can write that, but it would be an exact rewrite of safe_pointers :)